libctru/source/gdbhio.c

```diff
@@ -0,0 +1,397 @@
+{
+	memset(out, 0, sizeof(struct stat));
+	out->st_dev = _gdbHioImportScalar32(in->st_dev);
+	out->st_ino = _gdbHioImportScalar32(in->st_ino);
+	out->st_mode = _gdbHioImportFileMode(_gdbHioImportScalar32(in->st_dev));
+	out->st_nlink = _gdbHioImportScalar32(in->st_nlink);
+	out->st_uid = _gdbHioImportScalar32(in->st_uid);
+	out->st_gid = _gdbHioImportScalar32(in->st_gid);
+	out->st_rdev = _gdbHioImportScalar32(in->st_rdev);
+	out->st_size = (off_t)_gdbHioImportScalar64(in->st_size);
+	out->st_blksize = (blksize_t)_gdbHioImportScalar64(in->st_blksize);
+	out->st_blocks = (blkcnt_t)_gdbHioImportScalar64(in->st_blocks);
+	out->st_atime = _gdbHioImportScalar32(in->st_atime);
+	out->st_mtime = _gdbHioImportScalar32(in->st_mtime);
+	out->st_ctime = _gdbHioImportScalar32(in->st_ctime);
+}
+
+struct gdbhio_timeval {
+	gdbhio_time_t tv_sec;  /* second */
+	u64   tv_usec; /* microsecond */
+};
+
+static void _gdbHioImportStructTimeval(struct timeval *out, const struct gdbhio_timeval *in)
+{
+	out->tv_sec = _gdbHioImportScalar32(in->tv_sec);
+	out->tv_usec = _gdbHioImportScalar64(in->tv_usec);
+}
+
+static void _gdbHioSetErrno(int gdbErrno, bool ctrlC)
+{
+	errno = _gdbHioImportErrno(gdbErrno);
+	g_gdbHioWasInterruptedByCtrlC = ctrlC;
+}
+
+static int _gdbHioSendSyncRequestV(const char *name, const char *paramFormat, va_list args)
+{
+	PackedGdbHioRequest req = {{0}};
+	memcpy(req.magic, "GDB", 4);
+	strncpy(req.functionName, name, 16);
+	strncpy(req.paramFormat, paramFormat, 8);
+
+	u32 numStrs = 0;
+	const char *str;
+
+	for (u32 i = 0; i < 8 && paramFormat[i] != 0; i++) {
+		switch (paramFormat[i]) {
+			case 'i':
+			case 'I':
+				req.parameters[i] = va_arg(args, u32);
+				break;
+			case 'L':
+				req.parameters[i] = va_arg(args, u64);
+				break;
+			case 'p':
+				req.parameters[i] = va_arg(args, uintptr_t);
+				break;
+			case 's':
+				str = va_arg(args, const char *);
+				req.parameters[i] = (uintptr_t)str;
+				req.stringLengths[numStrs++] = strlen(str)+1;
+				break;
+			default:
+				break;
+		}
+	}
+
+	if (R_FAILED(svcOutputDebugString((const char *)&req, 0)) || req.paramFormat[0] != 0) {
+		errno = EPIPE;
+		g_gdbHioWasInterruptedByCtrlC = false;
+		return -1;
+	}
+
+	_gdbHioSetErrno(req.gdbErrno, req.ctrlC);
+	return req.retval;
+}
+
+static int _gdbHioSendSyncRequest(const char *name, const char *paramFormat, ...)
+{
+	int ret = 0;
+	va_list args;
+	va_start(args, paramFormat);
+	ret = _gdbHioSendSyncRequestV(name, paramFormat, args);
+	va_end(args);
+	return ret;
+}
+
+int gdbHioOpen(const char *pathname, int flags, mode_t mode)
+{
+	return _gdbHioSendSyncRequest("open", "siI", pathname, _gdbHioExportOpenFlags(flags), _gdbHioExportFileMode(mode));
+}
+
+int gdbHioClose(int fd)
+{
+	return _gdbHioSendSyncRequest("close", "i", fd);
+}
+
+int gdbHioRead(int fd, void *buf, unsigned int count)
+{
+	return _gdbHioSendSyncRequest("read", "ipI", fd, buf, count);
+}
+
+int gdbHioWrite(int fd, const void *buf, unsigned int count)
```

Comment:
Why not pass through `count` as `unsigned long` and use `L`?

Reply:
Same as above